packages/flutter/lib/src/services/restoration.dart

```diff
@@ -0,0 +1,981 @@
+    // There are three cases to consider:
+    // 1. Claiming an id that has already been claimed.
+    // 2. Claiming an id that doesn't yet exist in [_rawChildren].
+    // 3. Claiming an id that does exist in [_rawChildren] and hasn't been
+    //    claimed yet.
+    // If an id has already been claimed (case 1) the current owner may give up
+    // that id later this frame and it can be re-used. In anticipation of the
+    // previous owner's surrender of the id, we return an empty bucket for this
+    // new claim and check in [_debugAssertIntegrity] that at the end of the
+    // frame the old owner actually did surrendered the id.
+    // Case 2 also requires the creation of a new empty bucket.
+    // In Case 3 we create a new bucket wrapping the existing data in
+    // [_rawChildren].
+
+    // Case 1+2: Adopt and return an empty bucket.
+    if (_claimedChildren.containsKey(id) || !_rawChildren.containsKey(id.value)) {
+      final RestorationBucket child = RestorationBucket.empty(
+        debugOwner: debugOwner,
+        id: id,
+      );
+      adoptChild(child);
+      return child;
+    }
+
+    // Case 3: Return bucket wrapping the existing data.
+    assert(_rawChildren[id.value] != null);
+    final RestorationBucket child = RestorationBucket.child(
+      id: id,
+      parent: this,
+      debugOwner: debugOwner,
+    );
+    _claimedChildren[id] = child;
+    return child;
+  }
+
+  /// Adopts the provided `child` bucket.
+  ///
+  /// The `child` will be dropped from its old parent, if it had one.
+  ///
+  /// The `child` is stored under its [id] in this bucket. If this bucket
+  /// already contains a child bucket under the same id, the owner of that
+  /// existing bucket must give it up (e.g. by moving the child bucket to a
+  /// different parent or by disposing it) before the end of the current frame.
+  /// Otherwise an exception indicating the illegal use of duplicated
+  /// [RestorationId]s will trigger in debug mode.
+  ///
+  /// No-op if the provided bucket is already a child of this bucket.
+  void adoptChild(RestorationBucket child) {
+    assert(_debugAssertNotDisposed());
+    assert(child != null);
+    if (child._parent != this) {
+      child._parent?._removeChildData(child);
+      child._parent = this;
+      _addChildData(child);
+      if (child._manager != _manager) {
+        _recursivelyUpdateManager(child);
+      }
+    }
+    assert(child._parent == this);
+    assert(child._manager == _manager);
+  }
+
+  void _dropChild(RestorationBucket child) {
+    assert(child != null);
+    assert(child._parent == this);
+    _removeChildData(child);
+    child._parent = null;
+    if (child._manager != null) {
+      _updateManager(null);
+      _recursivelyUpdateManager(child);
+    }
+  }
+
+  bool _needsSerialization = false;
+  void _markNeedsSerialization() {
+    if (!_needsSerialization) {
+      _needsSerialization = true;
+      _manager?.scheduleSerializationFor(this);
```

Comment:
GitHub ate my last comment thread here.

Perhaps we could add a debug flag here that prints or warns somehow if serialization is being scheduled without a manager?

I don't want to eagerly add features, but this seems like something that could become confusing for users. Maybe it would help if I better understood how you could end up unhooked here.

Reply:
After talking this over with Dan, I agree that we should assert here for now that the manager is not null. If in the future we support that unhooking mechanism I was talking about we can revisit that decision.

Reply:
Since you're asserting it's not null, you can drop the `?` right?

Reply:
Oh, it might be null if it's decommissioned. I see.